// radius/dist/js/current/steps.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define('radius/steps', ['exports', 'lodash', './utils'], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require('lodash'), require('./utils'));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.lodash, global.utils);
    global.radiusSteps = mod.exports;
  }
})(this, function (exports, _lodash, _utils) {
  'use strict';

  Object.defineProperty(exports, "__esModule", {
    value: true
  });

  var _lodash2 = _interopRequireDefault(_lodash);

  var _utils2 = _interopRequireDefault(_utils);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  exports['default'] = function () {
    var steps = getSteps();
    var stepWrap = document.querySelector('[data-step-wrap]');
    var INACTIVE_CLASS = 'is-inactive';

    if (_lodash2['default'].isEmpty(steps)) {
      return;
    }

    setInactive(steps);

    _lodash2['default'].forEach(steps, function (current, index) {
      stepCount(current, index);

      _utils2['default'].addEventListener(current, 'click', function (event) {
        var el = event.target;
        var nextStepEl = current.nextElementSibling;

        if (!_utils2['default'].hasAttribute(el, 'data-step-advance') || _utils2['default'].hasAttribute(el, 'disabled')) {
          return;
        }

        if (el.tagName.toLowerCase() !== 'label') {
          event.preventDefault();
        }

        activateNextStep(nextStepEl);
      });
    });

    function createStepString(step, total) {
      var templateStr = stepWrap.getAttribute('data-step-text');
      var compiledTemp = _lodash2['default'].template(templateStr);

      return compiledTemp({ step: step, total: total });
    }

    function activateNextStep(nextEl) {
      var elHasClass = _utils2['default'].hasClass(nextEl, INACTIVE_CLASS);
      var event = new CustomEvent('stepActivated');

      if (elHasClass) {
        _utils2['default'].removeClass(nextEl, INACTIVE_CLASS);
        nextEl.dispatchEvent(event);
      }
    }

    function getSteps() {
      return _lodash2['default'].toArray(document.querySelectorAll('[data-step]'));
    }

    function stepCount(step, index) {
      var stepCountEl = step.querySelector('[data-step-count]');
      var stepNumber = index + 1;
      var totalSteps = steps.length;
      var stepString = createStepString(stepNumber, totalSteps);

      if (stepCountEl === null) {
        return;
      }

      _utils2['default'].textContent(stepCountEl, stepString);
    }

    function setInactive() {
      _lodash2['default'].forEach(_lodash2['default'].tail(steps), function (current) {
        _utils2['default'].addClass(current, INACTIVE_CLASS);
      });
    }
  };
});
